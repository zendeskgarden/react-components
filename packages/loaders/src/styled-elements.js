/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { retrieveTheme } from '@zendeskgarden/react-theming';

export const DotsCircle = styled.circle.attrs(props => ({
  cx: 9,
  cy: 9,
  r: 9,
  transform: props.transform
}))``;

DotsCircle.propTypes = {
  transform: PropTypes.string
};

export const SpinnerCircle = styled.circle.attrs(props => ({
  cx: 40,
  cy: 40,
  r: 34,
  fill: 'none',
  stroke: 'currentColor',
  strokeLinecap: 'round',
  strokeWidth: props.strokeWidthValue,
  strokeDasharray: `${props.dasharrayValue} 250`,
  transform: props.transform
}))``;

SpinnerCircle.propTypes = {
  dasharrayValue: PropTypes.number,
  strokeWidthValue: PropTypes.number,
  transform: PropTypes.string
};

const StyledSvg = styled.svg.attrs(props => ({
  'data-garden-version': PACKAGE_VERSION,
  xmlns: 'http://www.w3.org/2000/svg',
  width: props.width,
  height: props.height,
  focusable: 'false',
  viewBox: `0 0 ${props.width} ${props.height}`,
  role: 'progressbar'
}))`
  width: 1em;
  height: 0.9em;
  color: ${props => props.color || 'inherit'};
  font-size: ${props => props.fontSize || 'inherit'};

  ${props => retrieveTheme(props.dataGardenId, props)};
`;

StyledSvg.propTypes = {
  color: PropTypes.string,
  fontSize: PropTypes.any
};

export const StyledSVG = React.forwardRef(
  ({ children, fontSize, width, height, ...other }, ref) => {
    return (
      <StyledSvg fontSize={fontSize} width={width} height={height} ref={ref} {...other}>
        {children}
      </StyledSvg>
    );
  }
);

StyledSVG.propTypes = {
  children: PropTypes.node,
  fontSize: PropTypes.any,
  height: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  width: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
};

export const LoadingPlaceholder = styled.div.attrs({
  role: 'progressbar'
})`
  display: inline;
  width: 1em;
  height: 0.9em;
  font-size: ${props => props.fontSize};
`;

LoadingPlaceholder.propTypes = {
  fontSize: PropTypes.any
};
