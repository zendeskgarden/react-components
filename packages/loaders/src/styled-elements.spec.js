/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React from 'react';
import { shallow, mount } from 'enzyme';
import { StyledCircle, StyledSVG, LoadingPlaceholder } from './styled-elements';

describe('styled-elements', () => {
  describe('StyledCircle', () => {
    it('renders correctly by default', () => {
      const wrapper = shallow(<StyledCircle transform="transform(2 3)" />);

      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('StyledSVG', () => {
    it('renders correctly by default', () => {
      const wrapper = mount(<StyledSVG />);

      expect(wrapper).toMatchSnapshot();
    });

    it('renders color correctly if provided', () => {
      const wrapper = mount(<StyledSVG color="red" />);

      expect(wrapper).toMatchSnapshot();
    });

    it('renders font-size correctly if provided', () => {
      const wrapper = mount(<StyledSVG fontSize="15px" />);

      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('LoadingPlaceholder', () => {
    it('renders correctly by default', () => {
      const wrapper = shallow(<LoadingPlaceholder />);

      expect(wrapper).toMatchSnapshot();
    });

    it('renders font-size correctly if provided', () => {
      const wrapper = shallow(<LoadingPlaceholder fontSize="15px" />);

      expect(wrapper).toMatchSnapshot();
    });
  });
});
