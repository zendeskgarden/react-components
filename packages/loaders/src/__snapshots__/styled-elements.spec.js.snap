// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styled-elements LoadingPlaceholder renders correctly by default 1`] = `
.c0 {
  display: inline;
  width: 1em;
  height: 0.9em;
}

<div
  className="c0"
  role="progressbar"
/>
`;

exports[`styled-elements LoadingPlaceholder renders font-size correctly if provided 1`] = `
.c0 {
  display: inline;
  width: 1em;
  height: 0.9em;
  font-size: 15px;
}

<div
  className="c0"
  fontSize="15px"
  role="progressbar"
/>
`;

exports[`styled-elements StyledCircle renders correctly by default 1`] = `
.c0 {
  will-change: transform;
}

<circle
  className="c0"
  cx={9}
  cy={9}
  r={9}
  transform="transform(2 3)"
/>
`;

exports[`styled-elements StyledSVG renders color correctly if provided 1`] = `
.c0 {
  width: 1em;
  height: 0.9em;
  color: red;
  font-size: inherit;
}

<StyledSVG
  color="red"
>
  <styled-elements__StyledSvg
    color="red"
  >
    <svg
      className="c0"
      color="red"
      data-garden-id="loaders.dots"
      data-garden-version="version"
      focusable="false"
      height={72}
      role="progressbar"
      viewBox="0 0 80 72"
      width={80}
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        fill="currentColor"
      />
    </svg>
  </styled-elements__StyledSvg>
</StyledSVG>
`;

exports[`styled-elements StyledSVG renders correctly by default 1`] = `
.c0 {
  width: 1em;
  height: 0.9em;
  color: inherit;
  font-size: inherit;
}

<StyledSVG>
  <styled-elements__StyledSvg>
    <svg
      className="c0"
      data-garden-id="loaders.dots"
      data-garden-version="version"
      focusable="false"
      height={72}
      role="progressbar"
      viewBox="0 0 80 72"
      width={80}
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        fill="currentColor"
      />
    </svg>
  </styled-elements__StyledSvg>
</StyledSVG>
`;

exports[`styled-elements StyledSVG renders font-size correctly if provided 1`] = `
.c0 {
  width: 1em;
  height: 0.9em;
  color: inherit;
  font-size: 15px;
}

<StyledSVG
  fontSize="15px"
>
  <styled-elements__StyledSvg
    fontSize="15px"
  >
    <svg
      className="c0"
      data-garden-id="loaders.dots"
      data-garden-version="version"
      focusable="false"
      fontSize="15px"
      height={72}
      role="progressbar"
      viewBox="0 0 80 72"
      width={80}
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        fill="currentColor"
      />
    </svg>
  </styled-elements__StyledSvg>
</StyledSVG>
`;
